### PR TITLE
HDDS-4771. [FSO]S3MultiPart: Implement InitiateMultiPartUpload

### PR DESCRIPTION
**Pull request by rakeshadr**
## What changes were proposed in this pull request?

This task is to implement Initiate multipart upload for a specified key.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-4771

## How was this patch tested?

Added S3 multipart UT cases.

### CODE REVIEW

**Comment by rakeshadr** [comment]
@linyiqun , appreciate if you could review and give your feedback. Thanks in advance!

**Review by linyiqun — COMMENTED** [review]
@rakeshadr , I see original S3InitiateMultipartUploadRequest doesn't support FS semantic (not add parent key if it doesn't exist). So  it's still intend to implement V1 version here?

**Comment by rakeshadr** [comment]
> till intend to implement V1 version here?

Good point @linyiqun. Yes, V1 version will make 100% FS semantics for multi part request as well. IMHO, its good to follow same pattern for all the ObjectStore APIs.
Does it make sense to you?

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmMultipartKeyInfo.java [hunk 1]
```diff
@@ -35,6 +35,7 @@
   private final ReplicationType replicationType;
   private final ReplicationFactor replicationFactor;
   private TreeMap<Integer, PartKeyInfo> partKeyInfoList;
+  private long parentID;
```
Comment (linyiqun) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmMultipartKeyInfo.java:38]: Can we add some comments for parentID? Something like what's the meaning of this and when this  parentID will make sense.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileRequest.java [hunk 2]
```diff
@@ -510,6 +512,28 @@ public static void addToOpenFileTable(OMMetadataManager omMetadataMgr,
...
+  /**
+   * Adding omKeyInfo to open file table.
+   *
+   * @param omMetadataMgr OM Metadata Manager
+   * @param batchOp       batch of db operations
+   * @param omFileInfo    omKeyInfo
+   * @param uploadID      uploadID
+   * @throws IOException DB failure
+   */
+  public static void addToOpenFileTable(OMMetadataManager omMetadataMgr,
```
Comment (linyiqun) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileRequest.java:525]: I didn't see this method is used in current PR change. Suppose this should be used in S3InitiateMultipartUploadResponseV1#addToDBBatch, right?

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestOzoneClientMultipartUploadV1.java [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+   * Ozone is made active by setting OZONE_ENABLED = true
+   *
+   * @throws IOException
+   */
+  @BeforeClass
+  public static void init() throws Exception {
+    OzoneConfiguration conf = new OzoneConfiguration();
+    conf.setBoolean(OMConfigKeys.OZONE_OM_ENABLE_FILESYSTEM_PATHS, true);
+    conf.set(OMConfigKeys.OZONE_OM_LAYOUT_VERSION,
+            OMConfigKeys.OZONE_OM_LAYOUT_VERSION_V1);
```
Comment (linyiqun) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestOzoneClientMultipartUploadV1.java:74]: Can we reuse existed method TestOMRequestUtil#configureFSOptimizedPaths for this?

**Review by linyiqun — COMMENTED** [review]
Haven't taken a deep look for the unit tests, some minor comments below. Please have a check @rakeshadr .

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmMultipartKeyInfo.java (hunk 1 above)
Comment (rakeshadr) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmMultipartKeyInfo.java:61]: Sure, I have pushed the changes in new commit. I updated same comments in OmKeyInfo class as well.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileRequest.java (hunk 2 above)
Comment (rakeshadr) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileRequest.java]: Done!

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestOzoneClientMultipartUploadV1.java (hunk 3 above)
Comment (rakeshadr) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestOzoneClientMultipartUploadV1.java]: Done!

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/response/s3/multipart/TestS3InitiateMultipartUploadResponseV1.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+import java.util.UUID;
+
+/**
+ * Class tests S3 Initiate MPU response.
+ */
+public class TestS3InitiateMultipartUploadResponseV1
+    extends TestS3InitiateMultipartUploadResponse {
+
+  @Test
+  public void addDBToBatch() throws Exception {
```
Comment (linyiqun) [inline hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/response/s3/multipart/TestS3InitiateMultipartUploadResponseV1.java:39]: Can we rename this to testAddDBToBatch

**Review by linyiqun — COMMENTED** [review]
Looks great now, one nit for the unit test.

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/response/s3/multipart/TestS3InitiateMultipartUploadResponseV1.java (hunk 4 above)
Comment (rakeshadr) [reply hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/response/s3/multipart/TestS3InitiateMultipartUploadResponseV1.java]: Sure, I've renamed it.

**Comment by rakeshadr** [comment]
> Looks great now, one nit for the unit test.

Thanks, Could you please approve the PR after final review.

**Review by linyiqun — APPROVED** [review]
>Thanks, Could you please approve the PR after final review.
I'm +1 for this.

**Comment by rakeshadr** [comment]
Thanks a lot @linyiqun for the useful review comments. I've merged the PR.

**Pull request merged**